src/transformers/benchmark/benchmark_utils_generic.py

```diff
@@ -0,0 +1,151 @@
+#                 if parameter_name == "self":
+#                     continue
+#                 elif parameter_name in arguments:
+#                     continue
+#                 if parameter_name in kwargs:
+#                     arguments[parameter_name] = kwargs[parameter_name]
+#                 else:
+#                     arguments[parameter_name] = keyword_parameters[parameter_name]
+#
+#             self.init_kwargs = arguments
+#
+#         func(*args, **kwargs)
+#
+#     return wrapper
+
+
+class BenchMark:
+    def __init__(self, *arg, **kwargs):
+        pass
+
+    def measure(self, func, **measure_kwargs):
+        raise NotImplementedError
+
+    def target(self, **target_kwargs):
+        raise NotImplementedError
+
+    def inputs(self, **inputs_kwargs):
+        return {}
+
+    def _convert_to_json(self, report):
+        if isinstance(report, list):
+            return [self._convert_to_json(x) for x in report]
+        if isinstance(report, dict):
+            return {k: self._convert_to_json(v) for k, v in report.items()}
+        if isinstance(report, type):
+            return report.__name__
+        return report
+
+    def report(self, result, run_info, only_result=False, output_path=None):
+        report = {"result": result}
+        if not only_result:
+            report["init_kwargs"] = getattr(self, "init_kwargs", {})
+            report["run_kwargs"] = run_info
+
+        report = self._convert_to_json(report)
+        if output_path is not None:
+            with open(output_path, "w", encoding="UTF-8") as fp:
+                json.dump(report, fp, ensure_ascii=False, indent=4)
+
+        return report
+
+    def run(self, measure_kwargs=None, target_kwargs=None, inputs_kwargs=None, report_kwargs=None):
+        if measure_kwargs is None:
+            measure_kwargs = {}
+        if target_kwargs is None:
+            target_kwargs = {}
+        if inputs_kwargs is None:
+            inputs_kwargs = {}
+        if report_kwargs is None:
+            report_kwargs = {}
+        if measure_kwargs is None:
+            measure_kwargs = {}
+
+        target = self.target(**target_kwargs)
+
+        all_inputs_kwargs = [inputs_kwargs] if isinstance(inputs_kwargs, dict) else inputs_kwargs
+        results = []
+        for _inputs_kwargs in all_inputs_kwargs:
+            inputs = self.inputs(**_inputs_kwargs)
+            result = self.measure(target, **measure_kwargs)(**inputs)
+            results.append(result)
+
+        run_info = {
+            "measure_kwargs": measure_kwargs,
+            "target_kwargs": target_kwargs,
+            "inputs_kwargs": inputs_kwargs,
+            "report_kwargs": report_kwargs,
+        }
+
+        if isinstance(inputs_kwargs, dict):
+            results = results[0]
+
+        return self.report(results, run_info, **report_kwargs)
+
+
+class SpeedBenchMark(BenchMark):
+    def measure(self, func, number=3, repeat=1):
+        def wrapper():
```

Comment:
```suggestion
    def measure(self, func, number=3, repeat=1):
    """ It’s tempting to calculate mean and standard deviation from the result vector and report these. However, this is not very useful. In a typical case, the lowest value gives a lower bound for how fast your machine can run the given code snippet; higher values in the result vector are typically not caused by variability in Python’s speed, but by other processes interfering with your timing accuracy. So the [min()](https://docs.python.org/2/library/functions.html#min) of the result is probably the only number you should be interested in. After that, you should look at the entire vector and apply common sense rather than statistics. From https://docs.python.org/2/library/timeit.html#timeit.Timer.repeat """ 
        def wrapper():
```

Reply:
interesting!